dat/generated_tables.py

```diff
@@ -1,76 +1,158 @@
-from typing import Sequence
-
-from dat.model.row_collections import RowCollection
-from dat.model.table import GeneratedReferenceTable, ReferenceTable
-
-reference_table_1 = GeneratedReferenceTable(
-    table_name='reference_table_1',
-    table_description='My first table',
-    column_names=['letter', 'number', 'a_float'],
-    partition_keys=[],
-    reader_protocol_version=2,
-    writer_protocol_version=2,
-    row_collections=[
-        RowCollection(
-            write_mode='overwrite',
-            rows=[
-                ('a', 1, 1.1),
-                ('b', 2, 2.2),
-                ('c', 3, 3.3),
-            ],
-        ),
-        RowCollection(
-            write_mode='append',
-            rows=[
-                ('d', 4, 4.4),
-                ('e', 5, 5.5),
-            ],
-        ),
-    ],
+import os
+from typing import Callable, List, Optional, Tuple
+
+from dat.models import ExpectedMetadata, TableMetadata
+from dat.spark_builder import create_spark_session
+
+spark = create_spark_session()
+
+registered_reference_tables: List[Tuple[TableMetadata, Callable]] = []
+
+
+def save_expected(
+        metadata: TableMetadata, version: Optional[int] = None) -> None:
+    '''Save the specified version of a Delta Table as a Parquet file.'''
+    query = spark.read.format('delta')
+    if version:
+        query = query.option('versionAsOf', version)
+    df = query.load(metadata.delta_root)
+
+    # Need to ensure directory exists first
+    os.makedirs(metadata.expected_root(version))
+
```

Comment:
This pull request flatten three modules that were intended to provide evolvability and composability. Wouldn't this lead to create a big python script that doesn't provide easy extension/modification?

Separation of concerns and single responsibility principles that were strong pillars for the previous code disappeared here, and I personally think these are good qualities that have to be preserved

Reply:
> Wouldn't this lead to create a big python script that doesn't provide easy extension/modification?

Hi @edmondop, fair concern. What about this script makes it not easily modified? It seems we just add another table creation method and use the `@reference_table` tag.

It seems to introduce a very low overhead to add new reference tables. Furrthermore, these tables can be completely customized (e.g. create version 0-15, but then delete 0-10.jsons). But you're the code principles expert so I'd love to hear what specific pain points you think it will introduce.

Reply:
So this code does multiple things: it reads the right version of the table,  it writes the data to file, and it also generates the metadata, and write it to file. 

Given the number of responsibilities that this function has, testing it is not trivial. One first improvement of the structure is possible by moving behaviour into helper functions, and letting ```save_expected``` only invoke children functions:

```python
def create_expected_metadata(table_metadata, version) -> ExpectedMetadata:
     pass

def read_data_from_delta_as_version(table_metadata:TableMetadata, version) -> pyspark.sql.Dataframe:
     pass

def  save_data_and_metadata(df, table_metadata, expected_metadata):
      pass

def save_expected(table_metadata, version)):
      df = read_data_from_delta_as_version(table_metadata, version)
       expected_metadata = create_expected_metadata(table_metadata,version)
      save_data_and_metadata(df, table_metadata, expected_metadata)   
```

If we look beyond how we organize code into functions, but also how we organize functions in modules, there are good arguments to claim that `save_expected` should live at `dat` level, while `create_expected_metadata` is code that could live in the `model`, and `read_data_from_delta_as_version` and `save_data_and_metadata` should live in the `io` module (probably the `writers` module should be renamed `io` modules)